flang/lib/Lower/Bridge.cpp

```diff
@@ -819,6 +807,30 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     builder->create<fir::SelectOp>(loc, selectExpr, indexList, blockList);
   }
 
+  /// Collect DO CONCURRENT or FORALL loop control information.
+  IncrementLoopNestInfo getConcurrentControl(
+      const Fortran::parser::ConcurrentHeader &header,
+      const std::list<Fortran::parser::LocalitySpec> &localityList = {}) {
+    IncrementLoopNestInfo incrementLoopNestInfo;
+    for (const auto &control :
+         std::get<std::list<Fortran::parser::ConcurrentControl>>(header.t))
+      incrementLoopNestInfo.emplace_back(
+          *std::get<0>(control.t).symbol, std::get<1>(control.t),
+          std::get<2>(control.t), std::get<3>(control.t), true);
+    auto &info = incrementLoopNestInfo.back();
+    info.maskExpr = Fortran::semantics::GetExpr(
+        std::get<std::optional<Fortran::parser::ScalarLogicalExpr>>(header.t));
+    for (const auto &x : localityList) {
+      if (const auto *localInitList =
+              std::get_if<Fortran::parser::LocalitySpec::LocalInit>(&x.u))
+        for (const auto &x : localInitList->v)
+          info.localInitSymList.push_back(x.symbol);
+      if (std::get_if<Fortran::parser::LocalitySpec::Local>(&x.u))
+        llvm_unreachable("do concurrent locality specs not implemented");
```

Comment:
Use `TODO("message")` here.

Reply:
There is a distinction. Use `llvm_unreachable` when there is an "impossible" condition that could happen because of an upstream regression, etc. That's better than assuming the "impossible" will never happen and blindly going along compiling garbage.

TODO should be used for holes in the code; things that must be done, but aren't done yet.